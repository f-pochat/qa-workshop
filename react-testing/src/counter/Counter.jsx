import React, { useCallback } from 'react'
import { Button } from '@/components/ui/button'
import { Input } from '@/components/ui/input'
import { useCounter } from './hooks'
import PropTypes from 'prop-types'

export const Counter = ({ initialValue, ...props }) => {
  const { count, decrement, reset, increment, setCount } = useCounter(initialValue)

  const setCounter = useCallback((e) => {
    setCount(e.target.value)
  }, [setCount])

  return (
    <div className='flex flex-col items-center gap-2' {...props}>
      <Input
        placeholder='Count'
        type='number'
        onChange={setCounter}
        value={count}
      />
      <div className='flex items-center justify-between w-full'>
        <Button onClick={decrement}>-</Button>
        <Button onClick={reset}>Reset</Button>
        <Button onClick={increment}>+</Button>
      </div>
    </div>
  )
}

Counter.propTypes = {
  initialValue: PropTypes.number
}

Counter.defaultProps = {
  initialValue: 0
}
