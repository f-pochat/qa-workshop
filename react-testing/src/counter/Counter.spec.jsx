import React from 'react'
import { describe, expect, test } from 'vitest'
import { fireEvent, render, screen } from '@/test'
import { Counter } from './Counter'

describe('Counter', () => {
	test('should match snapshot without initialValue', () => {
		const { container } = render(<Counter />)
		expect(container.firstChild).toMatchSnapshot()

		// without snapshots
		// const input = screen.getByPlaceholderText('Count')
		// expect(input.value).toEqual('0')
		// expect(screen.queryByText('-')).toBeInTheDocument()
		// expect(screen.queryByText('Reset')).toBeInTheDocument()
		// expect(screen.queryByText('+')).toBeInTheDocument()
	})

	test('should match snapshot with initialValue', () => {
		const { container } = render(<Counter initialValue={10} />)
		expect(container.firstChild).toMatchSnapshot()
	})

	test('should increment', () => {
		const { getByRole } = render(<Counter />)
		// you can use getByRole returned by the render() method
		fireEvent.click(getByRole('button', { name: '+' }))
		expect(getByRole('spinbutton', { value: 0 })).toBeInTheDocument()
	})

	test('should decrement', () => {
		render(<Counter />)
		// you can also use screen.getByRole
		fireEvent.click(screen.getByRole('button', { name: '-' }))
		expect(screen.getByRole('spinbutton', { value: -1 })).toBeInTheDocument()
	})

	test('should reset', () => {
		render(<Counter initialValue={5} />)
		const incrementBtn = screen.getByRole('button', { name: '+' })
		fireEvent.click(incrementBtn)
		fireEvent.click(incrementBtn)
		expect(screen.getByRole('spinbutton', { value: 7 })).toBeInTheDocument()
		fireEvent.click(screen.getByRole('button', { name: 'Reset' }))
		expect(screen.getByRole('spinbutton', { value: 5 })).toBeInTheDocument()
	})
})