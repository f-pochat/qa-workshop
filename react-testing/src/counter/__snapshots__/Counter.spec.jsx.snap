// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Counter > should match snapshot with initialValue 1`] = `
<div
  class="flex flex-col items-center gap-2"
>
  <input
    class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
    placeholder="Count"
    type="number"
    value="10"
  />
  <div
    class="flex items-center justify-between w-full"
  >
    <button
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
    >
      -
    </button>
    <button
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
    >
      Reset
    </button>
    <button
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
    >
      +
    </button>
  </div>
</div>
`;

exports[`Counter > should match snapshot without initialValue 1`] = `
<div
  class="flex flex-col items-center gap-2"
>
  <input
    class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
    placeholder="Count"
    type="number"
    value="0"
  />
  <div
    class="flex items-center justify-between w-full"
  >
    <button
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
    >
      -
    </button>
    <button
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
    >
      Reset
    </button>
    <button
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
    >
      +
    </button>
  </div>
</div>
`;
